### PR TITLE
Team driving support

### PR DESCRIPTION
**Pull request by pnc**
This implements the design from opencab#17 in the sample
consumer and provider.

At a high level, it provides a way for:

1. Identity providers to indicate when a second (or third, etc.) driver
   has logged into the device.
2. HOS providers to return HOS information for each logged-in user,
   including duty status and hours-of-service clocks.

It aims to be backwards compatible, so that single-driver functionality
will continue to work between an older consumer app and a newer provider
or vice-versa.

### CODE REVIEW

**Comment by ZonarTimH** [comment]
There will be two Zonar reviewers on this.  My team handles the IdentityContract portion.  I asked the devs to double-check my read.  They came back with a thumbs-up.  Zonar is fine with the IdentityContract piece of the change.

I can't speak for the HOSContract piece, which is handled by another team.  I've forwarded this to Quynh Nguyen, who manages that team.  His signoff is also needed before Zonar has fully approved.

LMK if you need anything further from Zonar!

**Review by ZonarTimH — COMMENTED** [review]
I said this in MR comments.  Reposting in case this routes to a different place:

At Zonar, IdentityContract and HOSContract are implemented by two different teams.  I manage the folks who implement IdentityContract.  We have reviewed the IdentityContract change, and we approve it.

The HOSContract implementers are managed by Quynh Nguyen.  I forwarded this to him at the beginning of the week.  Hopefully you get word from him soon.

Either way, please keep us all posted on when you want this change to go to the world.  Zonar will need to figure out the timing of a Shell release which implements these changes.  Thanks!

**Comment by pnc** [comment]
Thanks a ton @ZonarTimH!

Fixing the docs so they can be easily browsed is still on my list.

File: OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java [hunk 1]
```diff
@@ -91,6 +97,78 @@ private void logEvent(String event) {
...
+                            if (result != null) {
+                                Log.d(LOG_TAG, "Got result!");
+                                result.setClassLoader(HOSContract.class.getClassLoader());
+
+                                Version maxSupportedVersion = new Version("0.3");
+                                Version resultVersion;
+                                if (result.containsKey(HOSContract.KEY_VERSION) && result.get(HOSContract.KEY_VERSION) != null) {
+                                    resultVersion = new Version(result.getString(HOSContract.KEY_VERSION));
+                                } else {
+                                    resultVersion = new Version("0.3");
```
Comment (pnc) [inline OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:134]: I think this should be `0.2` per 4.3.4 but I might be missing something!

File: OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java [hunk 2]
```diff
@@ -106,7 +184,7 @@ private void callVehicleInformationProvider() {
                             SimpleDateFormat s = new SimpleDateFormat("MM/dd hh:mm:ss");
                             String dateTime = s.format(new Date());
                             try {
-                                result = resolver.call(authority, "getVehicleInformation", "0.2", null);
+                                result = resolver.call(authority, VehicleInformationContract.METHOD_GET_VEHICLE_INFORMATION, VehicleInformationContract.VERSION, null);
                             } catch (Exception ex) {
                                 Log.i(LOG_TAG, "Error calling provider: ", ex);
                                 adapterVehicleInformation.insert(dateTime + " : " + "Package: " + provider.packageName + ", Error: " + ex.getMessage(), 0);
```
Comment (pnc) [inline OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:196]: This should implement a version check per 4.3.2.

File: OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java [hunk 3]
```diff
@@ -161,18 +239,27 @@ private void callIdentityProviderGetLoginCredentials() {
...
                                 Log.d(LOG_TAG, "Got result!");
                                 result.setClassLoader(IdentityContract.class.getClassLoader());
-                                if (result.containsKey(IdentityContract.KEY_LOGIN_CREDENTIALS)) {
-                                    IdentityContract.LoginCredentials loginCredentials = result.getParcelable(IdentityContract.KEY_LOGIN_CREDENTIALS);
-                                    Log.d(LOG_TAG, "Got token: " + loginCredentials);
-                                    if (loginCredentials != null) {
-                                        adapterLoginCredentials.insert(dateTime + " : " + "Package: " + provider.packageName  + ", Token: " + loginCredentials.getToken() + ", Provider: " + loginCredentials.getProvider(), 0);
-                                    } else {
-                                        adapterLoginCredentials.insert(dateTime + " : " + "Package: " + provider.packageName + ", Error: No login credentials found", 0);
+                                if (result.containsKey(IdentityContract.KEY_LOGIN_CREDENTIALS) || result.containsKey(IdentityContract.KEY_ALL_LOGIN_CREDENTIALS)) {
```
Comment (pnc) [inline OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:248]: This should implement a version check per 4.3.2.

File: OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java [hunk 4]
```diff
@@ -214,7 +301,9 @@ private void callIdentityProviderGetActiveDrivers() {
                                     List<IdentityContract.Driver> drivers = result.getParcelableArrayList(IdentityContract.KEY_ACTIVE_DRIVERS);
                                     Log.d(LOG_TAG, "Got drivers: " + drivers);
                                     if (drivers != null && drivers.size() > 0) {
-                                        adapterActiveDrivers.insert(dateTime + " : " + "Package: " + provider.packageName + ", Driver Name: " + drivers.get(0).getUsername() + ", Is Driving: " + drivers.get(0).isDriving(), 0);
```
Comment (pnc) [inline OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:217]: This should implement a version check per 4.3.2.

**Comment by pnc** [comment]
@Mehantp @BoKlassen Please see the latest commit, which fixes the docs for the HOS "manage action" and "logout action" to specify that it's a URI, not an "intent string." (Android seems to have pretty completely eradicated the old `intent:` URI scheme for security reasons, so I don't think it's a good idea to pin the spec to that.)

https://github.com/opencabstandard/opencab/pull/21/commits/ae1a215989eab7a2374a9554dcc8d2d5e4d93b8f

**Comment by BoKlassen** [comment]
> @Mehantp @BoKlassen Please see the latest commit, which fixes the docs for the HOS "manage action" and "logout action" to specify that it's a URI, not an "intent string." (Android seems to have pretty completely eradicated the old `intent:` URI scheme for security reasons, so I don't think it's a good idea to pin the spec to that.)
> 
> [ae1a215](https://github.com/opencabstandard/opencab/commit/ae1a215989eab7a2374a9554dcc8d2d5e4d93b8f)

Awesome, thanks. This works better for us.

**Comment by pnc** [comment]
Good deal, that matches real-world usage of that property today, so it should be net no change for @ZonarTimH et al.

It does leave open the possibility of users seeing a disambiguation dialog if two apps try to use the register for the same URI scheme in their manifests, but I think that risk is fairly remote.

File: OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java [hunk 5]
```diff
@@ -145,9 +146,38 @@
...
+ * <code>com.example</code> defines a receiver called <code>com.example.IdentityChangedReceiver</code> and package
+ * <code>net.subtle.foo.bar.IdentityChangedReceiver</code>, a provider implementing the identity contract and
+ * broadcasting the <code>ACTION_IDENTITY_INFORMATION_CHANGED</code> event would send intents to both receiver classes.</li>
+ * <li>MUST publish events for each discovered receiver class by calling <code>sendBroadcast</code> with an explicit {@link android.content.Intent}
+ * with its component <code>property</code> set to the package and fully qualified class
+ * name of the enumerated receiver class. This ensures that apps which have been terminated, stopped, or updated
+ * since last launch will be started by the OS and given an opportunity to handle the event.</li>
+ * </ol>
+ *
+ * <h4>3.5. Backwards- and forwards-compatibility</h4>
```
Comment (pnc) [inline OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: @ZonarTimH @BoKlassen @Mehantp We've added a bunch of info here about version interop. Our intention is to clarify but not make big changes/cause rework for your teams.

Does everything here track with your vision and/or implementations? (In particular, we had to update our code and the provider samples to better handle requests from newer consumers.)

**Review by pnc — COMMENTED** [review]
Okay, sorry for the commit noise but I think this should be ready for merge.

@BoKlassen and @ZonarTimH, in the absence of a formal governance board meeting or something, would you both be willing to approve if this all checks out from your side?

File: OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java (hunk 5 above)
Comment (ZonarTimH) [reply OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: I'll get that double-checked w the team. Please note that there are two approvers at Zonar, myself for IdentityContract and Quynh Nguyen for HOSContract.  I think it'd be good for you to get Quynh in here directly, vs me alerting him when there's something to review.

Comment (pnc) [reply OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: @ZonarTimH Can do. Do you have Quynh's GitHub username? I can add to the reviewers list!

Comment (ZonarTimH) [reply OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: Sure, just checked.  He is @quynh_nguyen.  Thanks!

Comment (ZonarTimH) [reply OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: Pardon me, Quỳnh prefers his khongtu username.  I think I gave you gitlab above.  Go with khongtu.

Comment (ZonarTimH) [reply OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: Deleted my earlier comment so there's no confusion about usernames.

Comment (pnc) [reply OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: Thanks! Due to the magic of email notifications, I missed the deletion but have now invited both usernames to the repository's review/working group.

Comment (ZonarTimH) [reply OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:164]: I got that double-check with the team.  Sorry on the holdup.  The result is a green light.  Zonar is OK with the revised IdentityContract changes.

**Comment by pnc** [comment]
This resolves #17.

**Pull request merged**